### PR TITLE
Refactor transorm register

### PR DESCRIPTION
**Pull request by ccssu**
- [ ] - import tools 多进程问题
- [x] - oneflow_compile 添加 model.state dict() 转换的钩子
- [ ] - from functools import partial 解决

### CODE REVIEW

File: src/onediff/infer_compiler/import_tools/finder.py [hunk 1]
```diff
@@ -7,8 +9,30 @@
 
 
 PREFIX = "mock_"
-SUFFIX = ""
+SUFFIX = "_1f"
```
Comment (ccssu) [inline src/onediff/infer_compiler/import_tools/finder.py:12]: 加个后缀，不和 packages/diffusers/src/onediff_mock_extension/`mock_diffusers` 有名字重复。

**Comment by lixiang007666** [comment]
发现这个分支和最新的 Comfui master 组合，好像会触发下面的 bug：
<details>
<summary></summary>
<pre><code>

ERROR:root:!!! Exception during processing !!!
ERROR:root:Traceback (most recent call last):
  File "/data/home/lixiang/ComfyUI/execution.py", line 153, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
  File "/data/home/lixiang/ComfyUI/execution.py", line 83, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
  File "/data/home/lixiang/ComfyUI/execution.py", line 76, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
  File "/data/home/lixiang/ComfyUI/nodes.py", line 1237, in sample
    return common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
  File "/data/home/lixiang/ComfyUI/nodes.py", line 1207, in common_ksampler
    samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
  File "/data/home/lixiang/ComfyUI/comfy/sample.py", line 100, in sample
    samples = sampler.sample(noise, positive_copy, negative_copy, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 728, in sample
    return sample(self.model, noise, positive, negative, cfg, self.device, sampler(), sigmas, self.model_options, latent_image=latent_image, denoise_mask=denoise_mask, callback=callback, disable_pbar=disable_pbar, seed=seed)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 633, in sample
    samples = sampler.sample(model_wrap, sigmas, extra_args, callback, noise, latent_image, denoise_mask, disable_pbar)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 589, in sample
    samples = getattr(k_diffusion_sampling, "sample_{}".format(sampler_name))(model_k, noise, sigmas, extra_args=extra_args, callback=k_callback, disable=disable_pbar, **extra_options)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/data/home/lixiang/ComfyUI/comfy/k_diffusion/sampling.py", line 137, in sample_euler
    denoised = model(x, sigma_hat * s_in, **extra_args)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 287, in forward
    out = self.inner_model(x, sigma, cond=cond, uncond=uncond, cond_scale=cond_scale, model_options=model_options, seed=seed)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/data/home/lixiang/ComfyUI/comfy/k_diffusion/external.py", line 129, in forward
    eps = self.get_eps(input * c_in, self.sigma_to_t(sigma), **kwargs)
  File "/data/home/lixiang/ComfyUI/comfy/k_diffusion/external.py", line 155, in get_eps
    return self.inner_model.apply_model(*args, **kwargs)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 275, in apply_model
    out = sampling_function(self.inner_model.apply_model, x, timestep, uncond, cond, cond_scale, model_options=model_options, seed=seed)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 253, in sampling_function
    cond, uncond = calc_cond_uncond_batch(model_function, cond, uncond, x, timestep, max_total_area, model_options)
  File "/data/home/lixiang/ComfyUI/comfy/samplers.py", line 229, in calc_cond_uncond_batch
    output = model_function(input_x, timestep_, **c).chunk(batch_chunks)
  File "/data/home/lixiang/ComfyUI/comfy/model_base.py", line 66, in apply_model
    return self.diffusion_model(xc, t, context=context, control=control, transformer_options=transformer_options, **extra_conds).float()
  File "/data/home/lixiang/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 260, in __call__
    return self._oneflow_module(*args, **kwargs)
  File "/data/home/lixiang/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 124, in __call__
    output = dpl_graph(*mapped_args, **mapped_kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/graph.py", line 289, in __call__
    return self._dynamic_input_graph_cache(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/cache.py", line 115, in __call__
    return graph(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/graph.py", line 292, in __call__
    self._compile(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/graph.py", line 860, in _compile
    return self._compile_new(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/graph.py", line 879, in _compile_new
    _, eager_outputs = self.build_graph(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/graph.py", line 1421, in build_graph
    outputs = self.__build_graph(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/graph.py", line 1549, in __build_graph
    outputs = self.build(*lazy_args, **lazy_kwargs)
  File "/data/home/lixiang/onediff/src/onediff/infer_compiler/with_oneflow_compile.py", line 199, in build
    return self.model(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 186, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 236, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/diffusionmodules/openaimodel.py", line 619, in forward
    h = forward_timestep_embed(module, h, emb, context, transformer_options)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/diffusionmodules/openaimodel.py", line 56, in forward_timestep_embed
    x = layer(x, context, transformer_options)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 186, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 236, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/attention.py", line 573, in forward
    x = block(x, context=context[i], transformer_options=transformer_options)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 186, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 236, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/attention.py", line 403, in forward
    return checkpoint(self._forward, (x, context, transformer_options), self.parameters(), self.checkpoint)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/diffusionmodules/util.py", line 123, in checkpoint
    return func(*inputs)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/attention.py", line 468, in _forward
    n = self.attn1(n, context=context_attn1, value=value_attn1)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 186, in __call__
    result = self.__block_forward(*args, **kwargs)
  File "/data/home/lixiang/anaconda3/lib/python3.8/site-packages/oneflow/nn/graph/proxy.py", line 236, in __block_forward
    result = unbound_forward_of_module_instance(self, *args, **kwargs)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/attention.py", line 379, in forward
    out = optimized_attention(q, k, v, self.heads)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/ldm/modules/attention.py", line 161, in attention_sub_quad
    mem_free_total, mem_free_torch = model_management.get_free_memory(query.device, True)
  File "/data/home/lixiang/ComfyUI/mock_comfy_1f/model_management.py", line 574, in get_free_memory
    mem_free_torch = mem_reserved - mem_active
TypeError: unsupported operand type(s) for -: 'DummyModule' and 'DummyModule'

</code></pre>
</details>

File: src/onediff_mock_extension/mock_diffusers_quant/__init__.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+def convert_func(mod: flow.Tensor, verbose=False):
+    return mod
+
+
+ONEDIFF_CUSTOM_TORCH2OF_FUNC_TYPE_MAP = {
+    convert_func: flow.Tensor,
+}
+
+#         import diffusers_quant
```
Comment (strint) [inline src/onediff_mock_extension/mock_diffusers_quant/__init__.py:24]: 这些注释的删掉吧

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py [hunk 3]
```diff
@@ -1,109 +1,75 @@
-import oneflow as flow
+""" Custom node converters for torch2of.
+ONEDIFF_MODEL_CLASS_REPLACEMENT_MAP = { PYTORCH_MODEL_CLASS: ONEFLOW_MODEL_CLASS }
```
Comment (strint) [inline src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:2]: ONEDIFF_MODEL_CLASS_REPLACEMENT_MAP -> ONEDIFF_TORCH_TO_ONEFLOW_CLASS_MAP

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py [hunk 4]
```diff
@@ -1,109 +1,75 @@
...
 from ..import_tools import (
-    print_red,
-    print_green,
+    print_yellow,
     get_mock_cls_name,
+    import_module_from_path,
 )
-from .register import torch2of, default_converter
-from ._globals import update_class_proxies, _initial_package_names
+from .register import torch2of
```
Comment (strint) [inline src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:15]: torch2of -> torch2onef

of 比较有歧义，我们后面把 oneflow 简称为 onef 吧

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py [hunk 5]
```diff
@@ -1,109 +1,75 @@
...
-        from .mock_diffusers import AttnProcessor as mock_AttnProcessor
-        from diffusers.models.attention_processor import Attention
 
-        cls_key_value = {
-            get_mock_cls_name(Attention): mock_Attention,
-            get_mock_cls_name(
-                diffusers.models.attention_processor.AttnProcessor2_0
-            ): mock_AttnProcessor,
-        }
+def load_custom_node(module_path):
```
Comment (strint) [inline src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:19]: load_custom_type_map

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py [hunk 6]
```diff
@@ -1,109 +1,75 @@
...
-            print_red(f"Failed at importing customed comfy modules: {e=}")
-            raise e
+        print_yellow(f"Cannot import {module_path} module for custom register")
+        print_yellow(e)
 
 
-def import_diffusers_quant():
-    try:
+def _init_custom_register():
+    custom_nodes_dir = Path(__file__).parents[3] / "onediff_mock_extension"
```
Comment (strint) [inline src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:56]: 这里是约定了一个 path 和 file 么？

在使用侧主动调用 infer_compiler 下的注册接口，这样更好。而不是 infer_compiler 去找文件

File: src/onediff/infer_compiler/import_tools/finder.py (hunk 1 above)
Comment (strint) [reply src/onediff/infer_compiler/import_tools/finder.py:12]: 后缀使用 _onef 吧

File: src/onediff_mock_extension/mock_diffusers/__init__.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    Attention: Attention1f,
+    AttnProcessor2_0: AttnProcessor1f,
+}
+
+
+def convert_func(func):
+    return default_converter(func)
+
+
+ONEDIFF_CUSTOM_TORCH2OF_FUNC_TYPE_MAP = {convert_func: AttnProcessor2_0}
```
Comment (strint) [inline src/onediff_mock_extension/mock_diffusers/__init__.py:21]: 这里改为调用 infer_compiler 下的注册接口注册 type map 吧

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py (hunk 6 above)
Comment (ccssu) [reply src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:56]: 是约定了一个 默认 `path` 和一个 可以 `export ONEDIFF_CUSTOM_REGISTER_PATH=$ONEDIFF_CUSTOM_REGISTER_PATH:custom/path/pkgname` 

这里参考 ComfyUI写的 [【🔗】](
https://github.com/comfyanonymous/ComfyUI/blob/d2e27b48f169b0e5def6b9b2c7874e4010282921/nodes.py#L1737-L1751)

Comment (ccssu) [reply src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:56]: 在使用侧主动调用 infer_compiler 下的注册接口, 会有两个问题 1. 是一个自己安装的包(pip install diffusers) , 需要开始前主动调用。 2.自己写的拓展包，需要可以 import (`import pkg` )。

File: src/onediff_mock_extension/mock_diffusers/__init__.py (hunk 7 above)
Comment (ccssu) [reply src/onediff_mock_extension/mock_diffusers/__init__.py]: 已从 `mock_diffusers` 删除

File: src/onediff/infer_compiler/import_tools/finder.py (hunk 1 above)
Comment (hjchen2) [reply src/onediff/infer_compiler/import_tools/finder.py]: 最好不要简写，直接_oneflow

File: src/onediff/infer_compiler/with_oneflow_compile.py [hunk 8]
```diff
@@ -20,39 +19,58 @@ def __init__(self, torch_module, oneflow_module):
         self._torch_module = torch_module
         self._oneflow_module = oneflow_module
 
+    @property
+    def oneflow_module(self):
+        if self._oneflow_module is not None:
+            return self._oneflow_module
+        print("Compile torch module to oneflow module ...")
+        self._oneflow_module = torch2onef(self._torch_module)
+        return self._oneflow_module
```
Comment (ccssu) [inline src/onediff/infer_compiler/with_oneflow_compile.py:28]: 延时转换： 改为 调用 `__call__` ,`apply_model` , `decode`  才进行转换。

File: src/onediff/infer_compiler/with_oneflow_compile.py [hunk 9]
```diff
@@ -95,42 +89,44 @@ def get_graph(self):
...
-        return self.process_output(output)
+                output = self._deployable_module_model.oneflow_module(*args, **kwargs)
+        return output
+
+    def to(self, *args, **kwargs):
+        self._deployable_module_model.to(*args, **kwargs)
+        return self
 
     # TODO(): Just for transformers VAE decoder
+    @input_output_processor
```
Comment (ccssu) [inline src/onediff/infer_compiler/with_oneflow_compile.py:127]: 输入输出转换 封装到 `diffusers/src/onediff/infer_compiler/utils/args_tree_util.py`

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py (hunk 6 above)
Comment (jackalcooper) [reply src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:56]: 这个1和2都没看懂是什么意思，可以举个例子吗

File: src/onediff/infer_compiler/convert_torch_to_of/proxy.py [hunk 10]
```diff
@@ -33,7 +32,11 @@ def proxy_class(cls: type):
         return __of_mds[full_cls_name]
 
     raise RuntimeError(
-        f"can't find oneflow module for: {str(cls)} please register in custom_register.py!"
+        f"""
+        1. Replace can't find proxy oneflow module for: {str(cls)} \n 
+        2. Check if the initial package {_initial_package_names} contains the package where {str(cls)} is located, 
+        if not, you need to add it.
```
Comment (jackalcooper) [inline src/onediff/infer_compiler/convert_torch_to_of/proxy.py:38]: you need to add it.怎么加提示一下

File: src/onediff/infer_compiler/convert_torch_to_of/proxy.py [hunk 11]
```diff
@@ -33,7 +32,11 @@ def proxy_class(cls: type):
         return __of_mds[full_cls_name]
 
     raise RuntimeError(
-        f"can't find oneflow module for: {str(cls)} please register in custom_register.py!"
+        f"""
+        1. Replace can't find proxy oneflow module for: {str(cls)} \n
```
Comment (jackalcooper) [inline src/onediff/infer_compiler/convert_torch_to_of/proxy.py:36]: Replace can't find 为什么要加replace呢，这个replace是什么意思

File: src/onediff/infer_compiler/convert_torch_to_of/proxy.py [hunk 12]
```diff
@@ -48,8 +51,9 @@ def __getitem__(self, index):  # __getitem__
 
         if isinstance(self._1f_proxy_submod, Iterable):
             submod = self._1f_proxy_submod[index]
-            from .register import torch2of
-            return torch2of(submod)
+            from .register import torch2onef
+
+            return torch2onef(submod)
```
Comment (jackalcooper) [inline src/onediff/infer_compiler/convert_torch_to_of/proxy.py:56]: 为什么要改名 onef 呢，of，1f，或者oneflow，onef 目前都没有出现过，并且容易和onefx混淆

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py (hunk 6 above)
Comment (ccssu) [reply src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:56]: 1: 对于 `diffusers`的拓展需要按照以前写法重新写一遍。
https://github.com/Oneflow-Inc/diffusers/blob/80fe53462455e3eedb011b4cec7d27cffe52a923/src/onediff/infer_compiler/convert_torch_to_of/custom_register.py#L10-L33
2: 在使用侧主动调用 infer_compiler 下的注册接口: 对于 comfy的拓展，程序运行前要求确保可以 `from mock_comfy import CrossAttentionPytorch, SpatialTransformer`  ,并且按照以前写法重新写一遍。
https://github.com/Oneflow-Inc/diffusers/blob/80fe53462455e3eedb011b4cec7d27cffe52a923/src/onediff/infer_compiler/convert_torch_to_of/custom_register.py#L36-L68

File: src/onediff_mock_extension/mock_diffusers/__init__.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+
+from diffusers.models.attention_processor import Attention, AttnProcessor2_0
+
+from .attention_processor_1f import Attention as Attention1f
+from .attention_processor_1f import AttnProcessor as AttnProcessor1f
+
+ONEDIFF_TORCH_TO_ONEF_CLASS_MAP = {
+    Attention: Attention1f,
+    AttnProcessor2_0: AttnProcessor1f,
+}
```
Comment (strint) [inline src/onediff_mock_extension/mock_diffusers/__init__.py:12]: import infer_compiler
infer_compiler.type_map_registery.register(Attention: Attention1f)

File: src/onediff/infer_compiler/convert_torch_to_of/custom_register.py (hunk 6 above)
Comment (strint) [reply src/onediff/infer_compiler/convert_torch_to_of/custom_register.py:56]: 1、registry
- mock, args: type = path
- replace, args: type = cla / func
```
onediff.infer_compiler.registery(
```

2、compile
```
onediff.infer_compiler.oneflow_compile
```

**Comment by lixiang007666** [comment]
> 发现这个分支和最新的 Comfui master 组合，好像会触发下面的 bug：

image

File: src/infer_compiler_registry/register_diffusers/__init__.py [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+from .attention_processor_1f import Attention as Attention1f
+from .attention_processor_1f import AttnProcessor as AttnProcessor1f
+
+torch2of_class_map = {
+    Attention: Attention1f,
+    AttnProcessor2_0: AttnProcessor1f,
+}
+
+register(package_names=["diffusers"], torch2of_class_map=torch2of_class_map)
```
Comment (strint) [inline src/infer_compiler_registry/register_diffusers/__init__.py:15]: 提供了一个注册 package 和 class 的接口

File: src/infer_compiler_registry/register_diffusers_quant/quant_diffusion_pipeline.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+from diffusers import StableDiffusionXLPipeline
+import diffusers_quant
+from diffusers_quant.utils import (
+    rewrite_sdxl_pipeline_attention,
+    replace_sub_module_with_quantizable_module,
+)
+from torch._dynamo import allow_in_graph as maybe_allow_in_graph
+
+
+def _use_graph():
```
Comment (strint) [inline src/infer_compiler_registry/register_diffusers_quant/quant_diffusion_pipeline.py:13]: 这是是很早期的写法，现在应该没用了

File: src/onediff/infer_compiler/convert_torch_to_of/_globals.py [hunk 16]
```diff
@@ -1,50 +1,48 @@
...
     print_green,
-    print_red,
 )
 
-__all__ = [
-    "update_class_proxies",
-]
+__all__ = ["update_class_proxies", "load_class_proxies_from_packages"]
+_ONEDIFF_TORCH_TO_OF_CLASS_MAP = {}
+_ONEDIFF_LOADED_PACKAGES = []
```
Comment (strint) [inline src/onediff/infer_compiler/convert_torch_to_of/_globals.py:15]: 有变量，有方法，这种情况写成一个类吧

**Review by strint — APPROVED** [review]

**Pull request merged**